cloud/pkg/operationcontroller/controller/util.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2022 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"strings"
+
+	metav1 "k8s.io/api/core/v1"
+
+	"github.com/kubeedge/kubeedge/pkg/apis/operations/v1alpha1"
+)
+
+const (
+	UpgradeStatusKey   = "upgrade.kubeedge.io/status"
+	UpgradeStatusValue = ""
+	UpgradeIDKey       = "upgradeid.kubeedge.io/%d"
+)
+
+// filterVersion returns true only if the edge node version already on the upgrade req
+// version is like: v1.22.6-kubeedge-v1.10.0-beta.0.185+95378fb019912a, expected is like v1.10.0
+func filterVersion(version string, expected string) bool {
+	// if not correct version format, also return true
+	index := strings.Index(version, "-kubeedge-")
+	if index == -1 {
+		return false
+	}
+
+	if version[index+10:] != expected {
+		// filter nodes that already in the required version
+		return false
+	}
+
+	return true
+}
+
+// isEdgeNode checks whether a node is an Edge Node
+// only if label {"node-role.kubernetes.io/edge": ""} exists, it is an edge node
+func isEdgeNode(node *metav1.Node) bool {
+	if node.Labels == nil {
+		return false
+	}
+	if _, ok := node.Labels["node-role.kubernetes.io/edge"]; !ok {
+		return false
+	}
+
+	if node.Labels["node-role.kubernetes.io/edge"] != "" {
+		return false
+	}
+
+	return true
+}
+
+// isCompleted returns true only if some/all edge upgrade is upgrading or completed
+func isCompleted(upgrade *v1alpha1.NodeUpgrade) bool {
+	// all edge node upgrade is upgrading or completed
+	if upgrade.Status.State != v1alpha1.InitialValue {
+		return true
+	}
+
+	// partial edge node upgrade is upgrading or completed
+	for _, status := range upgrade.Status.UpgradeStatus {
+		if status.UpgradeState != v1alpha1.InitialValue {
+			return true
+		}
+	}
+
+	return false
+}
+
+// RemoveDuplicateElement deduplicate
+func RemoveDuplicateElement(s []string) []string {
+	result := make([]string, 0, len(s))
+	temp := map[string]struct{}{}
```

Comment:
```suggestion
	temp := make(map[string]struct{}, len(s))
```

Reply:
all above fixed, please review again @vincentgoat 
cc @fisherxu @wackxu

Reply:
NodeUpgrade->NodeUpgradeJob, and keep only one upgrade history for one node
Upgrade is not allowed to modify once it's created
if upgrade successfully, insert one annotation to the node like `nodeupgradejob.operations.kubeedge.io/history: "v1.9.0->v1.10.0->v1.11.1"`, and keep the max and last 3 records